pkg/reconciler/testing/featured/featured.go

```diff
@@ -0,0 +1,361 @@
+/*
+Copyright 2021 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package featured
+
+import (
+	"bufio"
+	"context"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+	"testing"
+
+	"github.com/cucumber/godog"
+	"github.com/cucumber/godog/colors"
+	"github.com/cucumber/messages-go/v10"
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/go-cmp/cmp/cmpopts"
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	clientgotesting "k8s.io/client-go/testing"
+	. "knative.dev/discovery/pkg/reconciler/testing/v1alpha1"
+	pkgtest "knative.dev/pkg/reconciler/testing"
+	"knative.dev/reconciler-test/pkg/manifest"
+)
+
+var opt = godog.Options{
+	Output: colors.Colored(os.Stdout),
+}
+
+func Run(m *testing.M) {
+	if !flag.Parsed() {
+		flag.Parse()
+	}
+
+	if len(flag.Args()) > 0 {
+		opt.Paths = flag.Args()
+	} else {
+		opt.Paths = []string{
+			filepath.Join(callerPath(), "testdata/features/"),
+		}
+	}
+
+	format := "progress"
+	for _, arg := range os.Args[1:] {
+		if arg == "-test.v=true" { // go test transforms -v option
+			format = "pretty"
+			break
+		}
+	}
+
+	opt.Format = format
+
+	os.Exit(m.Run())
+}
+
+func InitializeTestSuite(ctx *godog.TestSuiteContext) {
+	ctx.BeforeSuite(func() {})
+}
+
+func TestReconcileKind(t *testing.T, kind string, factory pkgtest.Factory) {
+	status := godog.TestSuite{
+		Name:                 kind,
+		TestSuiteInitializer: InitializeTestSuite,
+		ScenarioInitializer: func(s *godog.ScenarioContext) {
+			ReconcileKindFeatureContext(t, s, kind, factory)
+		},
+		Options: &opt,
+	}.Run()
+
+	if status != 0 {
+		t.Fail()
+	}
+}
+
+func ReconcileKindFeatureContext(t *testing.T, s *godog.ScenarioContext, kind string, factory pkgtest.Factory) {
+	ctx := context.Background()
+
+	rt := &ReconcilerTest{
+		t: t,
+		row: pkgtest.TableRow{
+			Name: kind,
+			Ctx:  ctx,
+			//OtherTestData:           nil,
+			//Objects:                 nil,
+			//Key:                     "",
+			//WantErr:                 false,
+			//WantCreates:             nil,
+			//WantUpdates:             nil,
+			//WantStatusUpdates:       nil,
+			//WantDeletes:             nil,
+			//WantDeleteCollections:   nil,
+			//WantPatches:             nil,
+			//WantEvents:              nil,
+			//WithReactors:            nil,
```

Comment:
I really would like to get WithReactors to fail in a way that would allow us to test, say failed gets without it always being interpreted as NotFound. We have bunch of code that we can't test that way. Just jotting it down here to inspire you to fix that :)

Reply:
have any examples?

Reply:
I think something like this:
https://github.com/knative/eventing/blob/master/pkg/reconciler/subscription/subscription.go#L120

I have not been able to add a reactor to Get to resource that does not pass the "IsNotFound"